src/source/source_state.js

```diff
@@ -0,0 +1,74 @@
+// @flow
+const util = require('../util/util');
+const Tile = require('./tile');
+
+export type FeatureStates = {[feature_id: string]: {[key: string]: string | number | boolean }};
```

Comment:
Do we intend to only allow `string | number | boolean` in feature states? Aren't GeoJSON properties allowed to include `null`, arrays, and objects now?

Reply:
I think everything but `object` is OK to use. Nested objects are not supported in expressions, so it wouldn't be usable for runtime styling, and possibly cause confusion.

Reply:
@anandthakker Are nested objects supported in GeoJSON feature properties? I suspect the least confusing thing will be parity between GeoJSON feature properties and feature state.

Reply:
Yep, they're supported.  E.g., the following works:

```js
map.on('load', () => {
    map.addLayer({
        id: 'test',
        source: {
            type: 'geojson',
            data: {
                type: 'Feature',
                properties: {
                    x: { y: 'hello' }
                },
                geometry: {
                    type: 'Point',
                    coordinates: [0, 0]
                }
            }
        },
        type: 'symbol',
        layout: {
            'text-field': ["get", "y", ["get", "x"]]
        }
    })
})
```

Reply:
Although note that in `queryRenderedFeatures`, array or object property values currently get stringified.

Reply:
Thanks for the correction @anandthakker! In that case, feature states should accept all JSON types. 
For `queryRenderedFeatures`, this means that arrays and objects would be stringified in `properties` but not in `state`.

Reply:
Yeah, that's fine -- qRF _shouldn't_ stringify them, it's just a limitation of a current implementation detail: https://github.com/mapbox/mapbox-gl-js/issues/2434